test/test_systemd.rb

```diff
@@ -0,0 +1,39 @@
+require_relative 'helper'
+require "puma/sd_notify"
+
+class TestSystemd < Minitest::Test
+  def setup
+    ::Dir::Tmpname.create("puma_socket") do |sockaddr|
+      @sockaddr = sockaddr
+      @socket = Socket.new(:UNIX, :DGRAM, 0)
+      socket_ai = Addrinfo.unix(sockaddr)
+      @socket.bind(socket_ai)
+      ENV["NOTIFY_SOCKET"] = sockaddr
+    end
+  end
+
+  def teardown
+    @socket.close if @socket
+    File.unlink(@sockaddr) if @sockaddr
+    @socket = nil
+    @sockaddr = nil
+  end
+
+  def socket_message
+    @socket.recvfrom(10)[0]
+  end
+
+  def test_notify
+    count = Puma::SdNotify.ready
+    assert_equal(socket_message, "READY=1")
+    assert_equal(ENV["NOTIFY_SOCKET"], @sockaddr)
+    assert_equal(count, 7)
+
+    count = Puma::SdNotify.stopping
+    assert_equal(socket_message, "STOPPING=1")
+    assert_equal(ENV["NOTIFY_SOCKET"], @sockaddr)
+    assert_equal(count, 10)
+
+    refute Puma::SdNotify.watchdog?
+  end
```

Comment:
I think we should 1 additional full-up integration test, where you go through the whole process of starting/stopping a Puma server so we can cover the stuff in the Launcher.